topsites.py

```diff
@@ -0,0 +1,180 @@
+
+ATS_DATEFORMAT = "%Y-%m-%dT%H:%M:%S.{}Z"
+ATS_HASH_ALGORITHM = "HmacSHA256"
+
+# Regions/Countries to dump to topsites.db
+REGIONS = {
+    "": (10000, 1),  # Global
+    "US": (1000, 2),
+    "FR": (1000, 2),
+    "IN": (1000, 2),
+    "DE": (1000, 2),
+    "TW": (1000, 2),
+    "ID": (1000, 2),
+    "HK": (1000, 2),
+    "SG": (1000, 2),
+    "PL": (1000, 2),
+    "GB": (1000, 2),
+    "RU": (1000, 2),
+}
+
+# Cache parsed sites, change priority if raised
+topsites = {}
+
+engine = create_engine("sqlite:///" + os.path.join(
+    app.config["BASE_DIR"], "topsites.db"))
+Base = declarative_base()
+Session = sessionmaker(bind=engine)
+session = Session()
+
+
+class Site(Base):
+    '''Define a sqlalchemy base object for a alexa top site.'''
+    __tablename__ = "topsites"
+
+    url = Column(String, primary_key=True)
+    priority = Column(Integer)
+    country_code = Column(String)
+    ranking = Column(Integer)
+
+    def __init__(self, url, priority, country_code, ranking):
+        self.url = url
+        self.priority = priority
+        self.country_code = country_code
+        self.ranking = ranking
+
+Base.metadata.create_all(engine)
+
+
+class TopSites:
+    '''Used to query top site info from alexa API to Site object'''
+    def __init__(self, access_key, secret_key):
+        self.access_key = access_key
+        self.secret_key = secret_key
+
+    def query(self, begin, begin_priority):
+        '''Query alexa with given range, save to Site'''
+        query_string = self.build_query(begin, 100)
+        toSign = "GET\n" + ATS_SERVICE_HOST + "\n/\n" + query_string
+        signature = self.gen_sign(toSign, self.secret_key)
+        uri = "{base}{query}&{signature}".format(
+            base=ATS_AWS_BASE_URL,
+            query=query_string,
+            signature=urlencode({"Signature": signature}, "UTF-8"))
+
+        response = requests.get(uri)
+
+        if response.status_code == requests.codes.ok:
+            # See http://docs.aws.amazon.com/AlexaTopSites/latest/index.html?QUERY_QueryRequests.html  # nopep8
+            dom = parseString(response.content)
+            sites = dom.getElementsByTagName("aws:Site")
+            for site in sites:
+                url = self.get_text(
+                    site.getElementsByTagName("aws:DataUrl")[0].childNodes)
+                rank = int(self.get_text(
+                    site.getElementsByTagName("aws:Rank")[0].childNodes))
+
+                priority = begin_priority
+                if rank > 100 and rank <= 1000:
+                    priority = priority + 1
+                elif rank > 1000 and rank <= 10000:
+                    priority = priority + 2
+
+                if url in topsites:
+                    site_row = topsites[url]
+                    # Update priority is new is lower
+                    if site_row.priority > priority:
+                        site_row.priority = priority
+                        site_row.country_code = self.country_code
```

Comment:
It's a question from Sergiu, if we have a site is low priority in global, but high priority in tier 1 country, finally the priority should be higher.
I put url and Site object in topsites dict. If url already exists but with higher priority (from P3 to P2), then update the priority and country code, we will know why it's important to us.